substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1079 @@
+
+			for ticket in tickets {
+				debug!(target: LOG_TARGET, "Checking ring proof");
+
+				let Some(ticket_id_output) = ticket.signature.outputs.get(0) else {
+					debug!(target: LOG_TARGET, "Missing ticket vrf output from ring signature");
+					continue
+				};
+				let ticket_id_input =
+					vrf::ticket_id_input(&randomness, ticket.body.attempt_idx, epoch_idx);
+
+				// Check threshold constraint
+				let ticket_id = vrf::make_ticket_id(&ticket_id_input, &ticket_id_output);
+				if ticket_id >= ticket_threshold {
+					debug!(target: LOG_TARGET, "Ignoring ticket over threshold ({:032x} >= {:032x})", ticket_id, ticket_threshold);
+					continue
+				}
+
+				// Check for duplicates
+				if TicketsData::<T>::contains_key(ticket_id) {
+					debug!(target: LOG_TARGET, "Ignoring duplicate ticket ({:032x})", ticket_id);
+					continue
+				}
+
+				// Check ring signature
+				let sign_data = vrf::ticket_body_sign_data(&ticket.body, ticket_id_input);
+				if ticket.signature.ring_vrf_verify(&sign_data, &verifier) {
+					TicketsData::<T>::set(ticket_id, Some(ticket.body));
+					valid_tickets
+						.try_push(ticket_id)
+						.expect("Input segment has same length as bounded destination vector; qed");
+				} else {
+					debug!(target: LOG_TARGET, "Proof verification failure for ticket ({:032x})", ticket_id);
+				}
+			}
+
+			if !valid_tickets.is_empty() {
+				Self::append_tickets(valid_tickets);
+			}
+
+			Ok(Pays::No.into())
+		}
+
+		/// Plan an epoch configuration change.
+		///
+		/// The epoch configuration change is recorded and will be announced at the begining
+		/// of the next epoch together with next epoch authorities information.
+		/// In other words, the configuration will be enacted one epoch later.
+		///
+		/// Multiple calls to this method will replace any existing planned config change
+		/// that has not been enacted yet.
+		#[pallet::call_index(1)]
+		#[pallet::weight(T::WeightInfo::plan_config_change())]
+		pub fn plan_config_change(
+			origin: OriginFor<T>,
+			config: EpochConfiguration,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+
+			ensure!(
+				config.redundancy_factor != 0 && config.attempts_number != 0,
+				Error::<T>::InvalidConfiguration
+			);
+			PendingEpochConfigChange::<T>::put(config);
+			Ok(())
+		}
+	}
+
+	#[pallet::validate_unsigned]
+	impl<T: Config> ValidateUnsigned for Pallet<T> {
+		type Call = Call<T>;
+
+		fn validate_unsigned(source: TransactionSource, call: &Self::Call) -> TransactionValidity {
+			let Call::submit_tickets { tickets } = call else {
+				return InvalidTransaction::Call.into()
+			};
+
+			// Discard tickets not coming from the local node or that are not included in a block
+			if source == TransactionSource::External {
+				warn!(
+					target: LOG_TARGET,
+					"Rejecting unsigned `submit_tickets` transaction from external source",
+				);
+				return InvalidTransaction::BadSigner.into()
+			}
```

Comment:
How should you send your tickets, if you are not part of the active validator set? This will not work with the current implementation. Can we not use some zk proof that is taking all tickets and the next author set as input to find out if this was send by a valid authority?

Reply:
Yes, you are 100% right!

The unsigned `submit` is accepted only if is already in block or if the submitter is part of current validator set.

This will be part of the upcoming RFC about inbound and outbound interfaces. 

For this RFC, I have several proposals about this topic. Stuff like:
1. how the nodes which are part of the mixnet can submit the tickets.
2. are mixnet nodes validators or can be normal nodes?
   - If are validators then these should relay tickets which are from other nodes part of the next epoch set
     based on some criteria (e.g. function of epoch randomness). If they don't do that maybe a ticket submitter can provide a proof about this behavior... and maybe there is a punishment for who doesn't fulfill their duties as mixnodes?) 
3. should we also provide a signed version of "submit" to allow tickets submission for pseudo-anonymous tickets relayers (even not part of mixnet)? In this case if tickets are good we may think about zero out the weight. Otherwise pay the fee.

As this area is left a bit as an implementation detail there is a lot to discuss.
And soon there will be some RFC proposal for discussion.

For the moment the pallet keeps this a bit undefined :-/